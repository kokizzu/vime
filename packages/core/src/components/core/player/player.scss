vime-player {
  /**
   * @prop --vime-font-family: A custom font family to be used throughout the player.
   * 
   * @prop --vime-theme: A custom theme (color) to be used throughout the player. Any valid CSS 
   * `color` property should work (HEX, RGBA, HLS, ...).
   */

  --vime-font-family: 'Helvetica Neue', 'Segoe UI', Helvetica, Arial, sans-serif;

  box-sizing: border-box;
  direction: ltr;
  font-family: var(--vime-font-family);
  -moz-osx-font-smoothing: auto;
  -webkit-font-smoothing: subpixel-antialiased;
  -webkit-tap-highlight-color: transparent;
  font-variant-numeric: tabular-nums;
  font-weight: 500;
  line-height: 1.7;
  width: 100%;
  display: block;
  max-width: 100%;
  min-width: 414px;
  position: relative;
  text-shadow: none;
  outline: 0;
  transition: box-shadow 0.3s ease;

  *,
  *::after,
  *::before {
    box-sizing: inherit;
  }

  button {
    font: inherit;
    line-height: inherit;
    width: auto;

    &::-moz-focus-inner {
      border:0;
    }
  }

  // Fix 300ms delay
  a,
  button,
  input,
  label {
    touch-action: manipulation;
  }

  &.video {
    height: 0;
    background: #000;
    overflow: hidden;
  }

  &.fullscreen {
    background: #000;
    margin: 0;
    border-radius: 0;
    width: 100%;
    height: 100%;
    padding-bottom: 0 !important;
  }

  .blocker {
    position: absolute;
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;
    z-index: $z-index-blocker;
    display: inline-block;
  }
}
