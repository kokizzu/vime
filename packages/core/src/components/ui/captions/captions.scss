  vime-captions {
    /**
    * @prop --captions-text-color: The color of the captions text.
    * @prop --captions-font-size: Font size of captions when device screen size is <768px.
    * @prop --captions-font-size-medium: Font size of captions when device screen size is >768px.
    * @prop --captions-font-size-large: Font size of captions when device screen size is >992px.
    * @prop --captions-font-size-xlarge: Font size of captions when device screen size is >1200px.
    * @prop --captions-cue-bg-color: The background color of active cues.
    * @prop --captions-cue-border-radius: The border radius of active cues.
    * @prop --captions-cue-padding: The padding around active cues.
    */

    position: absolute;
    left: 0;
    bottom: 0;
    width: 100%;
    text-align: center;
    color: var(--captions-text-color,#fff);
    font-size: var(--captions-font-size, 14px);
    padding: $control-spacing;
    z-index: $z-index-captions;
    display: none;
    pointer-events: none;
    transition: transform 0.4s ease-in-out;

    &.enabled {
      display: inline-block;
    }

    &.hidden {
      display: none !important;
    }
    
    @media (min-width: $bp-md) {
      font-size: var(--captions-font-size-medium, 18px);
    }
      
    @media (min-width: $bp-lg) {
      font-size: var(--captions-font-size-large, 20px);
    }
    
    @media (min-width: $bp-xl) {
      font-size: var(--captions-font-size-xlarge, 22px);
    }
  
    & > span {
      display: inline-block;
      background: var(--captions-cue-bg-color, $color-gray-600);
      border-radius: var(--captions-cue-border-radius, 2px);
      box-decoration-break: clone;
      line-height: 185%;
      padding: var(--captions-cue-padding, 0.2em 0.5em);
      white-space: pre-wrap;
      pointer-events: none;
      
      // Firefox adds a <div> when using getCueAsHTML().
      :global(div) {
        display: inline;
      }
      
      &:empty {
        display: none;
      }
    }
  }
  