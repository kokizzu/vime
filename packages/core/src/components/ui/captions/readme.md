# vime-captions

Renders and displays VTT cues by hooking into the `textTracks` player property. This is a simple
implementation that can only handle rendering one text track, and one cue for the given track at a
time (even if many are active). The active track can be changed by setting the mode of any track
in the list to `showing`.

Be aware that after you set the text track mode to `showing`, the component will automatically set
it to hidden to avoid double captions. This also means that this component is **not recommended**
to be used in combination with the native HTML5 player controls.

## Visual

<img
  src="https://raw.githubusercontent.com/vime-js/vime/master/packages/core/src/components/ui/captions/captions.png"
  alt="Vime captions component"
/>

<!-- Auto Generated Below -->

## Usage

### Angular

```html {5} title="example.html"
<vime-player>
  <!-- ... -->
  <vime-ui>
    <!-- ... -->
    <vime-captions></vime-captions>
  </vime-ui>
</vime-player>
```

### Html

```html {5}
<vime-player>
  <!-- ... -->
  <vime-ui>
    <!-- ... -->
    <vime-captions></vime-captions>
  </vime-ui>
</vime-player>
```

### React

```tsx {2,10}
import React from 'react';
import { VimePlayer, VimeUi, VimeCaptions } from '@vime/react';

function Example() {
  return render(
    <VimePlayer>
      {/* ... */}
      <VimeUi>
        {/* ... */}
        <VimeCaptions />
      </VimeUi>
    </VimePlayer>
  );
}
```

### Vue

```html {6,12,18} title="example.vue"
<template>
  <VimePlayer>
    <!-- ... -->
    <VimeUi>
      <!-- ... -->
      <VimeCaptions />
    </VimeUi>
  </VimePlayer>
</template>

<script>
  import { VimePlayer, VimeUi, VimeCaptions } from '@vime/vue';

  export default {
    components: {
      VimePlayer,
      VimeUi,
      VimeCaptions,
    },
  };
</script>
```

## Properties

| Property         | Attribute         | Description                                                                                         | Type      | Default |
| ---------------- | ----------------- | --------------------------------------------------------------------------------------------------- | --------- | ------- |
| `controlsHeight` | `controls-height` | The height of any lower control bar in pixels so that the captions can reposition when it's active. | `number`  | `0`     |
| `hidden`         | `hidden`          | Whether the captions should be visible or not.                                                      | `boolean` | `false` |

## Events

| Event          | Description                                                                                                             | Type                                  |
| -------------- | ----------------------------------------------------------------------------------------------------------------------- | ------------------------------------- |
| `vCuesChange`  | Emitted when the active cues change. A cue is active when `currentTime >= cue.startTime && currentTime <= cue.endTime`. | `CustomEvent<TextTrackCue[]>`         |
| `vTrackChange` | Emitted when the current track changes.                                                                                 | `CustomEvent<TextTrack \| undefined>` |

## CSS Custom Properties

| Name                           | Description                                               |
| ------------------------------ | --------------------------------------------------------- |
| `--captions-cue-bg-color`      | The background color of active cues.                      |
| `--captions-cue-border-radius` | The border radius of active cues.                         |
| `--captions-cue-padding`       | The padding around active cues.                           |
| `--captions-font-size`         | Font size of captions when device screen size is <768px.  |
| `--captions-font-size-large`   | Font size of captions when device screen size is >992px.  |
| `--captions-font-size-medium`  | Font size of captions when device screen size is >768px.  |
| `--captions-font-size-xlarge`  | Font size of captions when device screen size is >1200px. |
| `--captions-text-color`        | The color of the captions text.                           |

## Dependencies

### Used by

- [vime-default-ui](../default-ui)

### Graph

```mermaid
graph TD;
  vime-default-ui --> vime-captions
  style vime-captions fill:#f9f,stroke:#333,stroke-width:4px
```

---

_Built with [StencilJS](https://stenciljs.com/)_
