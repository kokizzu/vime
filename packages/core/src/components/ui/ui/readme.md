# vime-ui

Simple container that holds a collection of user interface components.

The only important role this component really has is, avoiding overlapping custom UI with the
native iOS media player UI. Therefore, custom UI is only displayed on iOS if the `playsinline` prop
is `true`, and the player is not in fullscreen mode.

<!-- Auto Generated Below -->

## Usage

### Angular

```html {3-5} title="example.html"
<vime-player>
  <!-- ... -->
  <vime-ui>
    <!-- ... -->
  </vime-ui>
</vime-player>
```

### Html

```html {3-5}
<vime-player>
  <!-- ... -->
  <vime-ui>
    <!-- ... -->
  </vime-ui>
</vime-player>
```

### React

```tsx {2,8}
import React from 'react';
import { VimePlayer, VimeUi } from '@vime/react';

function Example() {
  return render(
    <VimePlayer>
      {/* ... */}
      <VimeUi>{/* ... */}</VimeUi>
    </VimePlayer>
  );
}
```

### Vue

```html {4-6,11,16} title="example.vue"
<template>
  <VimePlayer>
    <!-- ... -->
    <VimeUi>
      <!-- ... -->
    </VimeUi>
  </VimePlayer>
</template>

<script>
  import { VimePlayer, VimeUi } from '@vime/vue';

  export default {
    components: {
      VimePlayer,
      VimeUi,
    },
  };
</script>
```

## Dependencies

### Used by

- [vime-default-ui](../default-ui)

### Graph

```mermaid
graph TD;
  vime-default-ui --> vime-ui
  style vime-ui fill:#f9f,stroke:#333,stroke-width:4px
```

---

_Built with [StencilJS](https://stenciljs.com/)_
