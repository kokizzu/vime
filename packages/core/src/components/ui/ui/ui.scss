vime-ui {
  position: relative;
  pointer-events: none;
  z-index: $z-index-ui;
  width: 100%;

  &.hidden {
    display: none;
  }

  &.video {
    position: absolute;
    top: 0;
    left: 0;
    height: 100%;
  }
}