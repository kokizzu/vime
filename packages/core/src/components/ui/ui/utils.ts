import { getElement } from '@stencil/core';
import { HTMLStencilElement } from '@stencil/core/internal';

export const findUIRoot = (ref: any) => {
  let ui = getElement(ref);

  while (!(/^VIME-UI$/.test(ui?.nodeName))) {
    ui = ui.parentElement as HTMLStencilElement;
  }

  return ui as HTMLVimeUiElement;
};
