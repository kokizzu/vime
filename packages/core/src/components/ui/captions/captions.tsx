import {
  h, Component, Prop, Watch, Host, State, Event, EventEmitter,
} from '@stencil/core';
import { PlayerProps } from '../../core/player/PlayerProps';
import { withPlayerContext } from '../../core/player/PlayerContext';
import { Disposal } from '../../core/player/Disposal';
import { isUndefined } from '../../../utils/unit';
import { listen } from '../../../utils/dom';

@Component({
  tag: 'vime-captions',
  styleUrl: 'captions.scss',
})
export class Captions {
  private textTracksDisposal = new Disposal();

  private textTrackDisposal = new Disposal();

  private state = new Map<TextTrack, TextTrackMode>();

  @State() isEnabled = false;

  @State() activeTrack?: TextTrack;

  @Watch('activeTrack')
  onActiveTrackChange() {
    this.vTrackChange.emit(this.activeTrack);
  }

  @State() activeCues: TextTrackCue[] = [];

  @Watch('activeCues')
  onActiveCuesChange() {
    this.vCuesChange.emit(this.activeCues);
  }

  /**
   * Whether the captions should be visible or not.
   */
  @Prop() hidden = false;

  /**
   * The height of any lower control bar in pixels so that the captions can reposition when it's
   * active.
   */
  @Prop() controlsHeight = 0;

  /**
   * @internal
   */
  @Prop() isControlsActive: PlayerProps['isControlsActive'] = false;

  /**
   * @internal
   */
  @Prop() isVideoView: PlayerProps['isVideoView'] = false;

  /**
   * @internal
   */
  @Prop() playbackStarted: PlayerProps['playbackStarted'] = false;

  /**
   * @internal
   */
  @Prop() textTracks?: PlayerProps['textTracks'];

  /**
   * Emitted when the current track changes.
   */
  @Event({ bubbles: false }) vTrackChange!: EventEmitter<TextTrack | undefined>;

  /**
   * Emitted when the active cues change. A cue is active when
   * `currentTime >= cue.startTime && currentTime <= cue.endTime`.
   */
  @Event({ bubbles: false }) vCuesChange!: EventEmitter<TextTrackCue[]>;

  disconnectedCallback() {
    this.cleanup();
  }

  private cleanup() {
    this.state.clear();
    this.textTracksDisposal.empty();
    this.textTrackDisposal.empty();
  }

  private onCueChange() {
    this.activeCues = Array.from(this.activeTrack?.activeCues ?? []);
  }

  private onTrackChange() {
    this.activeCues = [];
    this.textTrackDisposal.empty();
    if (isUndefined(this.activeTrack)) return;
    this.textTrackDisposal.add(listen(this.activeTrack!, 'cuechange', this.onCueChange.bind(this)));
  }

  private findActiveTrack() {
    let activeTrack: TextTrack;

    Array.from(this.textTracks!).forEach((track) => {
      if (isUndefined(activeTrack) && (track.mode === 'showing')) {
        // eslint-disable-next-line no-param-reassign
        track.mode = 'hidden';
        activeTrack = track;
        this.state.set(track, 'hidden');
      } else {
        // eslint-disable-next-line no-param-reassign
        track.mode = 'disabled';
        this.state.set(track, 'disabled');
      }
    });

    return activeTrack!;
  }

  private onTracksChange() {
    let hasChanged = false;

    Array.from(this.textTracks!).forEach((track) => {
      if (!hasChanged) {
        hasChanged = !this.state.has(track) || (track.mode !== this.state.get(track));
      }

      this.state.set(track, track.mode);
    });

    if (hasChanged) {
      const activeTrack = this.findActiveTrack();
      if (this.activeTrack !== activeTrack) {
        this.activeTrack = activeTrack;
        this.onTrackChange();
      }
    }
  }

  @Watch('textTracks')
  onTextTracksListChange() {
    this.cleanup();
    if (isUndefined(this.textTracks)) return;
    this.onTracksChange();
    this.textTracksDisposal.add(listen(this.textTracks!, 'change', this.onTracksChange.bind(this)));
  }

  @Watch('isVideoView')
  @Watch('playbackStarted')
  onEnabledChange() {
    this.isEnabled = this.playbackStarted && this.isVideoView;
  }

  private renderCurrentCue() {
    const currentCue = this.activeCues[0];
    if (isUndefined(currentCue)) return '';
    const div = document.createElement('div');
    div.append((currentCue as VTTCue)!.getCueAsHTML());
    return div.innerHTML.trim();
  }

  render() {
    return (
      <Host
        style={{
          transform: `translateY(-${this.isControlsActive ? this.controlsHeight : 0}px)`,
        }}
        class={{
          enabled: this.isEnabled,
          hidden: this.hidden,
        }}
      >
        <span>{this.renderCurrentCue()}</span>
      </Host>
    );
  }
}

withPlayerContext(Captions, [
  'isVideoView',
  'playbackStarted',
  'isControlsActive',
  'textTracks',
]);
