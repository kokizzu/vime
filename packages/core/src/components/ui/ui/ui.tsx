import {
  h, Host, Component, Prop,
} from '@stencil/core';
import { withPlayerContext } from '../../core/player/PlayerContext';
import { PlayerProp, PlayerProps } from '../../core/player/PlayerProp';
import { IS_IOS } from '../../../utils/support';

/**
 * @slot - Used to pass in UI components for the player.
 */
@Component({
  tag: 'vime-ui',
  styleUrl: 'ui.scss',
})
export class UI {
  /**
   * @internal
   */
  @Prop() isVideoView: PlayerProps[PlayerProp.isVideoView] = false;

  /**
   * @internal
   */
  @Prop() playsinline: PlayerProps[PlayerProp.playsinline] = false;

  /**
   * @internal
   */
  @Prop() isFullscreenActive: PlayerProps[PlayerProp.isFullscreenActive] = false;

  render() {
    const canShowCustomUI = !IS_IOS
      || !this.isVideoView
      || (this.playsinline && !this.isFullscreenActive);

    return (
      <Host
        class={{
          hidden: !canShowCustomUI,
          video: this.isVideoView,
        }}
      >
        <div>
          {canShowCustomUI && <slot />}
        </div>
      </Host>
    );
  }
}

withPlayerContext(UI, [
  PlayerProp.isVideoView,
  PlayerProp.playsinline,
  PlayerProp.isFullscreenActive,
]);
