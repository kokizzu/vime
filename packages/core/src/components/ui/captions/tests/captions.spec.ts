import { SpecPage } from '@stencil/core/testing';
import { newUISpecPage } from '../../ui/tests';
import { PlayerProp } from '../../../core/player/PlayerProp';
import { ViewType } from '../../../core/player/ViewType';
import { Captions } from '../captions';

let page: SpecPage;
let provider: HTMLVimeFaketubeElement;
let captions: HTMLVimeCaptionsElement;

beforeEach(async () => {
  ({ page, provider } = await newUISpecPage(
    [Captions],
    '<vime-captions/>',
  ));

  captions = page.root!.querySelector('vime-captions')!;
});

it('should be structurally sound', () => {
  expect(captions).toMatchSnapshot();
});

it('should not render if not a video view', async () => {
  await provider.dispatchStateChange(PlayerProp.PlaybackStarted, true);
  await provider.dispatchStateChange(PlayerProp.ViewType, ViewType.Audio);
  await page.waitForChanges();
  expect(captions).not.toHaveClass('enabled');
});

it('should render if a video view and playback has started', async () => {
  await provider.dispatchStateChange(PlayerProp.PlaybackStarted, true);
  await provider.dispatchStateChange(PlayerProp.ViewType, ViewType.Video);
  await page.waitForChanges();
  expect(captions).toHaveClass('enabled');
});

it('should not be visible if hidden', async () => {
  captions.hidden = true;
  await page.waitForChanges();
  expect(captions).toHaveClass('hidden');
});

it('should adjust position based on controls height', async () => {
  await provider.dispatchStateChange(PlayerProp.IsControlsActive, true);
  captions.controlsHeight = 140;
  await page.waitForChanges();
  expect(captions.style.transform).toEqual('translateY(-140px)');
  await provider.dispatchStateChange(PlayerProp.IsControlsActive, false);
  await page.waitForChanges();
  expect(captions.style.transform).toEqual('translateY(-0px)');
});
