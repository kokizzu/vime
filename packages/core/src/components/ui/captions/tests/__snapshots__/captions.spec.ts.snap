// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should be structurally sound 1`] = `
<vime-captions style="transform: translateY(0);;">
  <span></span>
</vime-captions>
`;
